File: 227. Basic Calculator II.cpp
class Solution {
public:
    int calculate(string s) {
    int res = 0, num = 0, temp = 0;
    char op = '+';
    for (int i = 0; i < s.length(); i++)
    {
        if (isdigit(s[i]))
        {
            num = 0;
            while (i < s.length() && isdigit(s[i]))
            {
                num = num * 10 + s[i] - '0';
                i++;
            }
            if (op == '+' || op == '-')
            {
                res += temp;
                temp = num * (op == '-' ? -1 : 1);
            }
            else if (op == '*')
                temp *= num;
            else if (op == '/')
                temp /= num;
        }
        if (s[i] != ' ')
            op = s[i];
    }
    res += temp;
    return res;
    }
};

File: 227. Basic Calculator II.py
# timeout

class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = 0
        while i <len(s):
            if s[i] == u'+':
                s = s[:i] + " + " +s[i+1:]
                i += 2
            elif s[i] == u'-':
                s = s[:i] + " - " +s[i+1:]
                i += 2
            elif s[i] == u'*':
                s = s[:i] + " * " +s[i+1:]
                i += 2
            elif s[i] == u'/':
                s = s[:i] + " / " +s[i+1:]
                i += 2
            i += 1
        s = s.split(' ')
        num = []
        ope = []
        i = 0
        while i < len(s):
            if s[i] == u"*":
                n1 = num.pop()
                while s[i+1] == ' ' or s[i+1] == '':
                    i += 1
                num += [n1 * int(s[i+1])]
                i += 1
            elif s[i] == u"/":
                n1 = num.pop()
                while s[i+1] == ' ' or s[i+1] == '':
                    i += 1
                num += [n1 / int(s[i+1])]
                i += 1
            elif s[i] == u'+' or s[i] == u'-':
                ope += [s[i]]
            elif s[i] != ' ' and s[i] != '':
                num += [int(s[i])]
            i += 1
        res = num.pop(0)
        while num:
            n2 = num.pop(0)
            operator = ope.pop(0)
            if operator == u'+':
                res += n2
            else:
                res -= n2
        return res
